### DISCUSSION TITLE
Provenance Hash Question in Art Engine

### DISCUSSION THREAD

**Discussion by ghost** [nftchef/art-engine#225] [Help + Q&A] [answered]
So I have the provenance hash set to true in the config.js, run "npm run generate", and then run "node utils/provenance.js" in the terminal. 

The _provenance.json file is created successfully and the keccak-256 hash of the concatenatedHashString is correct, as I confirmed it by using this website: 
- https://emn178.github.io/online-tools/keccak_256.html

I am however a bit confused on how to verify and prove to anyone who potentially has questions as to how each individual hash is created in each metadata file. Is there anyway to prove this? The GitHub says "Then… After generating images and data, each metadata file will include an imageHash property, which is a Keccak256 hash of the output image." I have tried to verify on two seperate websites that convert image files to keccak-256 and got different results from what the art-engine created. Those websites are.
 - https://emn178.github.io/online-tools/keccak_256_checksum.html
 - https://hash.rfctools.com/keccak-256-hash-generator/

Is there any reason this should be happening / can someone explain to me better how the process for each individual hash is executed? Appreciate any help in advace!!

**Comment by nftchef** [comment, marked as answer]
hey @0xYorkie Thanks for bringing this up.

tl;dr, I've pushed a fix that solves this issue and allows _regular_ proof through tools like https://emn178.github.io/online-tools/keccak_256_checksum.html

https://github.com/nftchef/art-engine/commit/98a37b55fa5a85b3f37ed7243175aec2f83e3e57


## The issues prior to 2022/08/15

The generator was incorrectly hashing the image Buffer _after_ being JSON stringified. 
_previously_
```
const hashable = typeof input === Buffer ? input : JSON.stringify(input);
```
For anyone who needs to reverse engineer the hash prior to the latest change, you will need to run it through a script where 
A. read the file as a buffer, e.g. `readFileSync`
B. `JSON.stringify` the buffer, 
C. Return the hash hex
